src/LowerWarpShuffles.cpp

```diff
@@ -0,0 +1,760 @@
+        if (this_lane.defined() || op->name == "__shared") {
+            // Not a warp-level allocation
+            return IRMutator2::visit(op);
+        } else {
+            // Pick up this allocation and deposit it inside the loop over lanes at reduced size.
+            allocations.push_back(Stmt(op));
+            return mutate(op->body);
+        }
+    }
+
+public:
+    LowerWarpShuffles() {}
+};
+
+class HoistWarpShufflesFromSingleIfStmt : public IRMutator2 {
+    using IRMutator2::visit;
+
+    Scope<int> stored_to;
+    vector<pair<string, Expr>> lifted_lets;
+
+    Expr visit(const Call *op) override {
+        // If it was written outside this if clause but read inside of
+        // it, we need to hoist it.
+        if (starts_with(op->name, "llvm.nvvm.shfl.") &&
+            !expr_uses_vars(op, stored_to)) {
+            string name = unique_name('t');
+            lifted_lets.push_back({name, op});
+            return Variable::make(op->type, name);
+        } else {
+            return IRMutator2::visit(op);
+        }
+    }
+
+    template<typename ExprOrStmt, typename LetOrLetStmt>
+    ExprOrStmt visit_let(const LetOrLetStmt *op) {
+        Expr value = mutate(op->value);
+        ExprOrStmt body = mutate(op->body);
+
+        // If any of the lifted expressions use this, we also need to
+        // lift this.
+        bool should_lift = false;
+        for (const auto &p : lifted_lets) {
+            should_lift |= expr_uses_var(p.second, op->name);
+        }
+
+        if (should_lift) {
+            lifted_lets.push_back({op->name, value});
+            return body;
+        } else {
+            return LetOrLetStmt::make(op->name, value, body);
+        }
+    }
+
+    Expr visit(const Let *op) override {
+        return visit_let<Expr>(op);
+    }
+
+    Stmt visit(const LetStmt *op) override {
+        return visit_let<Stmt>(op);
+    }
+
+    Stmt visit(const Store *op) override {
+        stored_to.push(op->name, 0);
+        return IRMutator2::visit(op);
+    }
+public:
+    Stmt rewrap(Stmt s) {
+        while (!lifted_lets.empty()) {
+            const pair<string, Expr> &p = lifted_lets.back();
+            s = LetStmt::make(p.first, p.second, s);
+            lifted_lets.pop_back();
+        }
+        return s;
+    }
+};
+
+// The destination *and source* for warp shuffles must be active
+// threads, or the value is undefined, so we want to lift them out of
+// if statements.
+class HoistWarpShuffles : public IRMutator2 {
+    using IRMutator2::visit;
+
+    Stmt visit(const IfThenElse *op) override {
+        // Move all Exprs that contain a shuffle out of the body of
+        // the if.
+        Stmt then_case = mutate(op->then_case);
+        Stmt else_case = mutate(op->else_case);
+
+        HoistWarpShufflesFromSingleIfStmt hoister;
+        then_case = hoister.mutate(then_case);
+        else_case = hoister.mutate(else_case);
+        Stmt s = IfThenElse::make(op->condition, then_case, else_case);
```

Comment:
Most other mutators only make the new IR if something changes. Looks like it might be a pain to do that in this case though.